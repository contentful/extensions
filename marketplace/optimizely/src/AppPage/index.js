import React from 'react';
import PropTypes from 'prop-types';
import tokens from '@contentful/forma-36-tokens';
import { css } from 'emotion';
import { Heading, Typography, Paragraph } from '@contentful/forma-36-react-components';

import Connect from './Connect';
import Config from './Config';
import OptimizelyLogo from './OptimizelyLogo';
import SectionSplitter from '../EditorPage/subcomponents/section-splitter';
import { VARIATION_CONTAINER_ID } from './constants';

const styles = {
  body: css({
    margin: '0 auto',
    padding: '20px 40px',
    maxWidth: '786px',
    minHeight: '700px',
    marginTop: tokens.spacingXl,
    backgroundColor: '#fff',
    zIndex: '2',
    boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.1)',
    borderRadius: '2px'
  }),
  background: css({
    display: 'block',
    position: 'absolute',
    zIndex: '-1',
    top: '0',
    width: '100%',
    height: '300px',
    backgroundColor: '#bcc3ca'
  }),
  section: css({
    margin: `${tokens.spacingXl} 0`
  }),
  featuresListItem: css({
    listStyleType: 'disc',
    marginLeft: tokens.spacingM
  }),
  light: css({
    opacity: '0.6',
    marginTop: tokens.spacingM
  }),
  logo: css({
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    marginTop: tokens.spacingXl,
    marginBottom: tokens.spacingXl
  })
};

export default class AppPage extends React.Component {
  static propTypes = {
    openAuth: PropTypes.func.isRequired,
    accessToken: PropTypes.string,
    client: PropTypes.object,
    sdk: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      config: {
        optimizelyProjectId: '',
        contentTypes: {}
      },
      allContentTypes: []
    };
  }

  async componentDidMount() {
    const { app } = this.props.sdk.platformAlpha;

    const [currentParameters, { items: allContentTypes = [] }] = await Promise.all([
      app.getParameters(),
      this.props.sdk.space.getContentTypes()
    ]);

    const enabledContentTypes = this.findEnabledContentTypes(allContentTypes);

    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState(prevState => ({
      allContentTypes,
      config: {
        contentTypes: enabledContentTypes,
        optimizelyProjectId: currentParameters
          ? currentParameters.optimizelyProjectId
          : prevState.optimizelyProjectId
      }
    }));

    app.onConfigure(this.configureApp);
  }

  configureApp = async () => {
    if (!this.props.accessToken) {
      this.props.sdk.notifier.error(`You must be connected to Optimizely to configure the app.`);
      return false;
    }

    const { config } = this.state;

    if (!config.optimizelyProjectId) {
      this.props.sdk.notifier.error(
        'You must provide an Optimizely project id in order to run experiments!'
      );
      return false;
    }

    const needsVariationContainerInSpace = !this.state.allContentTypes.find(
      ct => ct.sys.id === VARIATION_CONTAINER_ID
    );

    if (needsVariationContainerInSpace) {
      await this.createVariationContainerContentType();
    }

    const res = await this.saveEnabledContentTypes(
      this.state.config.contentTypes,
      this.state.allContentTypes
    );

    this.props.sdk.space
      .getContentTypes()
      .then(data => this.setState({ allContentTypes: data.items }));

    if (!res) {
      this.props.sdk.notifier.error('Something went wrong, please try again.');
      return false;
    }

    return {
      parameters: {
        optimizelyProjectId: config.optimizelyProjectId
      },
      targetState: {
        EditorInterface: {
          [VARIATION_CONTAINER_ID]: { editor: true, sidebar: { position: 0 } }
        }
      }
    };
  };

  createVariationContainerContentType = async () => {
    const variationContainer = await this.props.sdk.space.createContentType({
      sys: {
        id: VARIATION_CONTAINER_ID
      },
      name: 'Variation Container',
      displayField: 'experimentTitle',
      fields: [
        {
          id: 'experimentTitle',
          name: 'Experiment title',
          type: 'Symbol'
        },
        {
          id: 'experimentId',
          name: 'Experiment ID',
          type: 'Symbol'
        },
        {
          id: 'meta',
          name: 'Meta',
          type: 'Object'
        },
        {
          id: 'variations',
          name: 'Variations',
          type: 'Array',
          items: {
            type: 'Link',
            validations: [],
            linkType: 'Entry'
          }
        },
        {
          id: 'experimentKey',
          name: 'Experiment Key',
          type: 'Symbol'
        }
      ]
    });

    await this.props.sdk.space.updateContentType(variationContainer);
  };

  saveEnabledContentTypes = async (contentTypes, allContentTypes) => {
    const copyAllCts = JSON.parse(JSON.stringify(allContentTypes));
    const output = [];

    for (const ct of copyAllCts) {
      let hasChanges = false;

      for (const contentField of ct.fields) {
        const validations =
          contentField.type === 'Array' ? contentField.items.validations : contentField.validations;
        const index = (validations || []).findIndex(v => v.linkContentType);

        if (index > -1) {
          const linkValidations = validations[index];
          const includesVariationContainer = linkValidations.linkContentType.includes(
            VARIATION_CONTAINER_ID
          );

          const fieldsToEnable = contentTypes[ct.sys.id] || {};

          if (!includesVariationContainer && fieldsToEnable[contentField.id]) {
            linkValidations.linkContentType.push(VARIATION_CONTAINER_ID);
            hasChanges = true;
          }

          if (
            includesVariationContainer &&
            (!Object.keys(contentTypes).includes(ct.sys.id) || !fieldsToEnable[contentField.id])
          ) {
            linkValidations.linkContentType = linkValidations.linkContentType.filter(
              lct => lct !== VARIATION_CONTAINER_ID
            );
            hasChanges = true;
          }
        }
      }

      if (hasChanges) {
        output.push(ct);
      }
    }

    if (!output.length) {
      return true;
    }

    const updates = output.map(ct => {
      return this.props.sdk.space.updateContentType(ct);
    });

    try {
      await Promise.all(updates);
      return true;
    } catch (e) {
      return false;
    }
  };

  findEnabledContentTypes = (allContentTypes = []) => {
    return allContentTypes.reduce((acc, ct) => {
      const output = {};

      for (const field of ct.fields) {
        if (field.type === 'Array' && field.items.linkType === 'Entry') {
          output[field.id] = field.items.validations.some(val =>
            val.linkContentType.includes(VARIATION_CONTAINER_ID)
          );
          continue;
        }

        if (field.type === 'Link' && field.linkType === 'Entry') {
          output[field.id] =
            field.validations.length === 0 ||
            field.validations.some(val => val.linkContentType.includes(VARIATION_CONTAINER_ID));
        }
      }

      const keys = Object.keys(output);

      if (keys.some(key => output[key])) {
        return { ...acc, [ct.sys.id]: output };
      }

      return acc;
    }, {});
  };

  updateConfig = config => {
    this.setState(state => ({
      config: {
        ...state.config,
        ...config
      }
    }));
  };

  render() {
    return (
      <>
        <div className={styles.background} />
        <div className={styles.body}>
          <div className={styles.section}>
            <Typography>
              <Heading>Connect Optimizely</Heading>
              {!this.props.client ? (
                <div className={styles.section}>
                  <Connect openAuth={this.props.openAuth} />
                </div>
              ) : (
                <>
                  <Paragraph>You&rsquo;re currently connected to Optimizely.</Paragraph>
                  <Paragraph className={styles.light}>
                    An access token is valid for 2 hours then you must reauthorize.
                  </Paragraph>
                </>
              )}
            </Typography>
          </div>
          {!!this.props.client && (
            <div className={styles.section}>
              <SectionSplitter />
              <Config
                client={this.props.client}
                config={this.state.config}
                updateConfig={this.updateConfig}
                allContentTypes={this.state.allContentTypes}
                sdk={this.props.sdk}
              />
            </div>
          )}
        </div>
        <div className={styles.logo}>
          <OptimizelyLogo />
        </div>
      </>
    );
  }
}
