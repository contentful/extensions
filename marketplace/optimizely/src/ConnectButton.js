import React from 'react';

export default function ConnectButton() {
  return (
    <svg id="Layer_1" data-name="Layer 1" width="40" height="40" viewBox="40 40">
      <path
        d="M25.5,12.77a7.65,7.65,0,0,0-1.91,3.28,7.31,7.31,0,0,0-5.41,7l5.15-5.15c0,.44,0,.89.09,1.38l-5.1,5.15a7.44,7.44,0,0,0,.75,2.17l3.28-3.28a3.24,3.24,0,0,1,.89-2.4,4.7,4.7,0,0,1,.53-.44h0a7.24,7.24,0,0,0,4.3,4.3h0a3.05,3.05,0,0,1-2.8,1.46L22,29.55a7.44,7.44,0,0,0,2.17.75l5.19-5.15a6.89,6.89,0,0,0,1.38.13l-5.19,5.1a7.31,7.31,0,0,0,7-5.41,7.28,7.28,0,1,0-3.77-14.07A7,7,0,0,0,25.5,12.77Zm7.37,7.37a3.11,3.11,0,1,1,0-4.44,3.19,3.19,0,0,1,0,4.44Z"
        style={{fill: '#fff'}}
      />
      <path
        d="M23.28,25.33a3.76,3.76,0,0,1-.62-.84l-3.06,3,.67.8c.27.27.53.49.8.71l3-3Z"
        style={{fill: '#fff'}}
      />
    </svg>
  );
}
