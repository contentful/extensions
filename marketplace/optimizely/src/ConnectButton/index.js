import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';
import { Button } from '@contentful/forma-36-react-components';

import OptimizelyLogo from './OptimizelyLogo';

const styles = {
  connect: css({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    '&:hover': {
      backgroundColor: 'transparent!important' // necessary to eliminate the forma styling in favor of the custom optimizely styling
    }
  })
};

export default function ConnectButton({ openAuth }) {
  return (
    <Button
      className={styles.connect}
      onClick={openAuth}
      testId="connect-button"
      isFullWidth
      buttonType="naked">
      <OptimizelyLogo />
    </Button>
  );
}

ConnectButton.propTypes = {
  openAuth: PropTypes.func.isRequired
};
