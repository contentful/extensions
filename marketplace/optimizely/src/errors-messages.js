import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';
import { Note } from '@contentful/forma-36-react-components';

const styles = {
  container: css({
    margin: tokens.spacingL
  })
};

const requiredFields = [
  {
    id: 'experimentTitle',
    type: 'Symbol'
  },
  { id: 'experimentId', type: 'Symbol' },
  { id: 'experimentKey', type: 'Symbol' },
  { id: 'meta', type: 'Object' },
  { id: 'variations', type: 'Array' }
];

export function isValidContentType(contentType) {
  const missing = [];

  requiredFields.forEach(item => {
    const exists = contentType.fields.find(
      field => field.id === item.id && field.type === item.type
    );
    if (!exists) {
      missing.push(item);
    }
  });

  return [missing.length === 0, missing];
}

export function MissingProjectId() {
  return (
    <div className={styles.container} data-test-id="missing-project">
      <Note noteType="negative">
        Please, set <strong>Optimizely Project ID</strong> in Extension settings.
      </Note>
    </div>
  );
}

export function IncorrectContentType(props) {
  const title = `Extension "${props.sdk.ids.extension}" cannot be assigned to this content type`;
  return (
    <div className={styles.container} data-test-id="incorrect-type">
      <Note noteType="negative" title={title}>
        <div>
          <strong>Required:</strong>{' '}
          {requiredFields.map(item => `${item.id} (${item.type})`).join(', ')}
        </div>
        <div>
          <strong>Missing:</strong>{' '}
          {props.missingFields.map(item => `${item.id} (${item.type})`).join(',')}
        </div>
      </Note>
    </div>
  );
}

IncorrectContentType.propTypes = {
  missingFields: PropTypes.array.isRequired,
  sdk: PropTypes.shape({
    ids: PropTypes.shape({
      extension: PropTypes.string
    }).isRequired
  }).isRequired
};
