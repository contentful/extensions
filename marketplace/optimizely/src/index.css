html,
body {
  margin: 0;
  padding: 0;
  border: 0;
  vertical-align: baseline;
  font-family: 'Avenir Next W01', -apple-system, BlinkMacSystemTypography, 'Segoe UI', Cantarell,
    Helvetica, Arial, sans-serif;
}

@font-face {
  font-family: 'Avenir Next W01';
  src: url('https://static.contentful.com/app/avenir-regular-normal-457e94a5b6.woff2')
    format('woff2');
  font-weight: 500;
  font-style: normal;
}
@font-face {
  font-family: 'Avenir Next W01';
  src: url('https://static.contentful.com/app/avenir-medium-normal-24932ad03d.woff2')
    format('woff2');
  font-weight: 600;
  font-style: normal;
}

@font-face {
  font-family: 'Avenir Next W01';
  src: url('https://static.contentful.com/app/avenir-demibold-normal-d42bfc911f.woff2')
    format('woff2');
  font-weight: 700;
  font-style: normal;
}
