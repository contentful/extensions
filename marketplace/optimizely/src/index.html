<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
  </head>
  <body>
    <div id="root"></div>
      <script>
        if (location.hash) {
          window.opener.postMessage({hash: location.hash});
          window.close();
        }
      </script>
    <script type="text/javascript" src="./index.js"></script>
  </body>
</html>
