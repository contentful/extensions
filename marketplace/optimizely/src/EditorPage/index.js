/* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events, jsx-a11y/anchor-is-valid */
import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';
import useMethods from 'use-methods';
import tokens from '@contentful/forma-36-tokens';
import { Note, Paragraph, Modal } from '@contentful/forma-36-react-components';
import StatusBar from './subcomponents/status-bar';
import ReferencesSection from './subcomponents/references-section';
import ExperimentSection from './subcomponents/experiment-section';
import VariationsSection from './subcomponents/variations-section';
import SectionSplitter from './subcomponents/section-splitter';
import { SDKContext, GlobalStateContext } from './subcomponents/all-context';
import prepareReferenceInfo, { COMBINED_LINK_VALIDATION_CONFLICT } from './reference-info';
import useInterval from '@use-it/interval';
import ConnectButton from '../ConnectButton';

const styles = {
  root: css({
    margin: tokens.spacingXl
  }),
  paragraph: css({
    marginBottom: tokens.spacingM
  }),
  link: css({
    cursor: 'pointer',
    textDecoration: 'underline'
  })
};

const methods = state => {
  return {
    setInitialData({ experiments, contentTypes, referenceInfo }) {
      state.experiments = experiments;
      state.contentTypes = contentTypes;
      state.referenceInfo = referenceInfo;
      state.loaded = true;
    },
    setError(message) {
      state.error = message;
    },
    setExperimentId(id) {
      state.experimentId = id;
      state.meta = {};
    },
    setVariations(variations) {
      state.variations = variations;
    },
    setEntry(id, entry) {
      state.entries[id] = entry;
    },
    setMeta(meta) {
      state.meta = meta;
    },
    setExperimentResults(id, results) {
      state.experimentsResults[id] = results;
    },
    updateExperiment(id, experiment) {
      const index = state.experiments.findIndex(
        experiment => experiment.id.toString() === id.toString()
      );
      if (index !== -1) {
        state.experiments[index] = experiment;
      }
    }
  };
};

const getInitialValue = sdk => ({
  loaded: false,
  error: false,
  experiments: [],
  contentTypes: [],
  meta: sdk.entry.fields.meta.getValue() || {},
  variations: sdk.entry.fields.variations.getValue() || [],
  experimentId: sdk.entry.fields.experimentId.getValue(),
  entries: {},
  experimentsResults: {}
});

const fetchInitialData = async (sdk, client) => {
  const { space, ids, locales } = sdk;

  const [contentTypesRes, entriesRes, experiments] = await Promise.all([
    space.getContentTypes(),
    space.getEntries({ links_to_entry: ids.entry, skip: 0, limit: 1000 }),
    client.getExperiments()
  ]);

  return {
    experiments,
    contentTypes: contentTypesRes.items,
    referenceInfo: prepareReferenceInfo({
      contentTypes: contentTypesRes.items,
      entries: entriesRes.items,
      variationContainerId: ids.entry,
      variationContainerContentTypeId: ids.contentType,
      defaultLocale: locales.default
    })
  };
};

function isCloseToExpiration(expires) {
  const _10minutes = 600000;
  return parseInt(expires, 10) - Date.now() <= _10minutes;
}

export default function EditorPage(props) {
  const globalState = useMethods(methods, getInitialValue(props.sdk));
  const [state, actions] = globalState;
  const [showAuth, setShowAuth] = useState(isCloseToExpiration(props.expires));

  const experiment = state.experiments.find(
    experiment => experiment.id.toString() === state.experimentId
  );

  /**
   * Fetch initial portion of data required to render initial state
   */
  useEffect(() => {
    fetchInitialData(props.sdk, props.client)
      .then(data => {
        actions.setInitialData(data);
        return data;
      })
      .catch(() => {
        actions.setError('Unable to load initial data');
      });
  }, [actions, props.client, props.sdk]);

  /**
   * Pulling current experiment every 5s to get new status and variations
   */
  useInterval(() => {
    if (state.experimentId) {
      props.client
        .getExperiment(state.experimentId)
        .then(experiment => {
          actions.updateExperiment(state.experimentId, experiment);
          return experiment;
        })
        .catch(() => {});
    }
  }, 5000);

  /*
   * Poll to see if we need to show the reauth flow preemptively
   */
  useInterval(() => {
    setShowAuth(isCloseToExpiration(props.expires));
  }, 5000);

  /**
   * Subscribe for changes in entry
   */
  useEffect(() => {
    const unsubsribeExperimentChange = props.sdk.entry.fields.experimentId.onValueChanged(data => {
      actions.setExperimentId(data);
    });
    const unsubsribeVariationsChange = props.sdk.entry.fields.variations.onValueChanged(data => {
      actions.setVariations(data || []);
    });
    const unsubscribeMetaChange = props.sdk.entry.fields.meta.onValueChanged(data => {
      actions.setMeta(data || {});
    });
    return () => {
      unsubsribeExperimentChange();
      unsubsribeVariationsChange();
      unsubscribeMetaChange();
    };
  }, [
    actions,
    props.sdk.entry.fields.experimentId,
    props.sdk.entry.fields.meta,
    props.sdk.entry.fields.variations
  ]);

  /**
   * Update title every time experiment is changed
   */
  useEffect(() => {
    if (state.loaded) {
      const title = experiment ? `[Optimizely] ${experiment.name}` : '';
      props.sdk.entry.fields.experimentTitle.setValue(title);
    }
  }, [experiment, props.sdk.entry.fields.experimentTitle, state.loaded]);

  /**
   * Fetch experiment results every time experiment is changed
   */
  useEffect(() => {
    if (state.loaded && experiment) {
      props.client
        .getExperimentResults(experiment.id)
        .then(results => {
          actions.setExperimentResults(experiment.id, results);
          return results;
        })
        .catch(() => {});
    }
  }, [actions, experiment, props.client, state.loaded]);

  const getExperimentResults = experiment => {
    if (!experiment) {
      return undefined;
    }
    return {
      url: props.client.getResultsUrl(experiment.campaign_id, experiment.id),
      results: state.experimentsResults[experiment.id]
    };
  };

  /**
   * Handlers
   */

  const onChangeExperiment = useCallback(value => {
    props.sdk.entry.fields.meta.setValue({});
    props.sdk.entry.fields.experimentId.setValue(value.experimentId);
    props.sdk.entry.fields.experimentKey.setValue(value.experimentKey);
  });

  const onLinkVariation = useCallback(async variation => {
    const data = await props.sdk.dialogs.selectSingleEntry({
      locale: props.sdk.locales.default,
      contentTypes: state.referenceInfo.linkContentTypes
    });

    if (!data) {
      return;
    }

    const values = props.sdk.entry.fields.variations.getValue() || [];
    const meta = props.sdk.entry.fields.meta.getValue() || {};
    props.sdk.entry.fields.meta.setValue({
      ...meta,
      [variation.key]: data.sys.id
    });
    props.sdk.entry.fields.variations.setValue([
      ...values,
      {
        sys: {
          type: 'Link',
          id: data.sys.id,
          linkType: 'Entry'
        }
      }
    ]);
  });

  const onOpenEntry = useCallback(entryId => {
    props.sdk.navigator.openEntry(entryId, { slideIn: true });
  });

  const onCreateVariation = useCallback(async (variation, contentTypeId) => {
    const data = await props.sdk.navigator.openNewEntry(contentTypeId, {
      slideIn: true
    });

    if (!data) {
      return;
    }

    const values = props.sdk.entry.fields.variations.getValue() || [];
    const meta = props.sdk.entry.fields.meta.getValue() || {};

    props.sdk.entry.fields.meta.setValue({
      ...meta,
      [variation.key]: data.entity.sys.id
    });
    props.sdk.entry.fields.variations.setValue([
      ...values,
      {
        sys: {
          type: 'Link',
          id: data.entity.sys.id,
          linkType: 'Entry'
        }
      }
    ]);
  });

  const onRemoveVariation = useCallback((entryId, variation) => {
    const values = props.sdk.entry.fields.variations.getValue() || [];
    const meta = props.sdk.entry.fields.meta.getValue() || {};
    if (variation) {
      delete meta[variation.key];
    }
    props.sdk.entry.fields.meta.setValue(meta);
    props.sdk.entry.fields.variations.setValue(values.filter(item => item.sys.id !== entryId));
  });

  const onClearVariations = useCallback(() => {
    props.sdk.entry.fields.meta.setValue({});
    props.sdk.entry.fields.variations.setValue([]);
  });

  const { combinedLinkValidationType } = state.referenceInfo || {};
  if (combinedLinkValidationType === COMBINED_LINK_VALIDATION_CONFLICT) {
    return (
      <Note noteType="negative" title="Conflict">
        Validations of reference fields in incoming references yield conflicting references for the
        Variation Container. Loosen validations or change incoming references so there is at least
        one shared Content Type validation.
      </Note>
    );
  }

  return (
    <SDKContext.Provider value={props.sdk}>
      <GlobalStateContext.Provider value={globalState}>
        <Modal title="Connect with Optimizely" isShown={!props.client}>
          <Paragraph className={styles.paragraph} testId="reconnect-optimizely">
            Your Optimizely session has expired. Reconnect to continue editing.
          </Paragraph>
          <ConnectButton openAuth={props.openAuth} />
        </Modal>
        <div className={styles.root} data-test-id="editor-page">
          <StatusBar
            loaded={state.loaded}
            experiment={experiment}
            variations={state.variations}
            entries={state.entries}
          />
          <SectionSplitter />
          {showAuth && (
            <Note noteType="warning" className={styles.paragraph}>
              Your Optimizely session will expire soon. Click here to{' '}
              <a onClick={props.openAuth} className={styles.link} data-test-id="preemptive-connect">
                connect with Optimizely.
              </a>
            </Note>
          )}
          <ReferencesSection
            loaded={state.loaded}
            references={state.loaded ? state.referenceInfo.references : []}
            sdk={props.sdk}
          />
          <SectionSplitter />
          <ExperimentSection
            loaded={state.loaded}
            disabled={experiment && state.variations.length > 0}
            experiments={state.experiments}
            experiment={experiment}
            onChangeExperiment={onChangeExperiment}
            onClearVariations={onClearVariations}
          />
          <SectionSplitter />
          <VariationsSection
            loaded={state.loaded}
            contentTypes={state.contentTypes}
            experiment={experiment}
            experimentResults={getExperimentResults(experiment)}
            meta={state.meta}
            variations={state.variations}
            onCreateVariation={onCreateVariation}
            onLinkVariation={onLinkVariation}
            onOpenEntry={onOpenEntry}
            onRemoveVariation={onRemoveVariation}
          />
        </div>
      </GlobalStateContext.Provider>
    </SDKContext.Provider>
  );
}

EditorPage.propTypes = {
  openAuth: PropTypes.func.isRequired,
  client: PropTypes.any.isRequired,
  expires: PropTypes.string.isRequired,
  sdk: PropTypes.shape({
    space: PropTypes.object.isRequired,
    ids: PropTypes.object.isRequired,
    locales: PropTypes.object.isRequired,
    navigator: PropTypes.shape({
      openEntry: PropTypes.func.isRequired,
      openNewEntry: PropTypes.func.isRequired
    }).isRequired,
    dialogs: PropTypes.shape({
      selectSingleEntry: PropTypes.func.isRequired
    }).isRequired,
    entry: PropTypes.shape({
      fields: PropTypes.shape({
        experimentId: PropTypes.object.isRequired,
        experimentKey: PropTypes.object.isRequired,
        variations: PropTypes.object.isRequired,
        meta: PropTypes.object.isRequired,
        experimentTitle: PropTypes.object.isRequired
      }).isRequired,
      getSys: PropTypes.func.isRequired
    }).isRequired,
    parameters: PropTypes.shape({
      installation: PropTypes.shape({
        optimizelyProjectId: PropTypes.string.isRequired
      }).isRequired
    }).isRequired
  }).isRequired
};
