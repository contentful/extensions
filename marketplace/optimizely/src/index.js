import React from 'react';
import PropTypes from 'prop-types';
import { render } from 'react-dom';
import { init, locations } from 'contentful-ui-extensions-sdk';
import './index.css';
import EditorPage from './EditorPage';
import Sidebar from './Sidebar';
import { IncorrectContentType, isValidContentType, MissingProjectId } from './errors-messages';
import OptimizelyClient from './optimizely-client';
import AppPage from './AppPage';
import '@contentful/forma-36-react-components/dist/styles.css';
import 'whatwg-fetch';

function parseHash(hash) {
  return hash
    .slice(1)
    .split('&')
    .reduce((acc, pair) => {
      const [key, value] = pair.split('=');
      acc[key] = value;
      return acc;
    }, {});
}

function getTokenExpirationTime(expires) {
  return Date.now() + parseInt(expires, 10) * 1000;
}

// we can use this for the oauth endpoint,
// if there is a hash with an access token, we will report it and close the page
if (window.location.hash) {
  const { access_token = null, expires_in = null } = parseHash(window.location.hash);
  const expireTime = getTokenExpirationTime(expires_in);

  window.opener.postMessage({ token: access_token, expires: expireTime });
  window.close();
}

const HOST = `${window.location.protocol}//${window.location.host}`;

const url = `https://app.optimizely.com/oauth2/authorize
?client_id=15687650042
&redirect_uri=${window.encodeURIComponent(HOST)}
&response_type=token
&scopes=all`;

const TOKEN_KEY = 'optToken';
const TOKEN_EXPIRATION = 'optExpire';

export default class App extends React.Component {
  static propTypes = {
    sdk: PropTypes.shape({
      contentType: PropTypes.object,
      location: PropTypes.shape({
        is: PropTypes.func.isRequired
      }),
      parameters: PropTypes.shape({
        installation: PropTypes.shape({
          optimizelyProjectId: PropTypes.string.isRequired
        }).isRequired
      }).isRequired
    }).isRequired
  };

  constructor(props) {
    super(props);

    const token = window.localStorage.getItem(TOKEN_KEY);
    const expires = window.localStorage.getItem(TOKEN_EXPIRATION);

    this.state = {
      client: token ? this.makeClient(token) : null,
      accessToken: token,
      expires
    };

    this.listener = window.addEventListener(
      'message',
      event => {
        const { data, origin } = event;
        const { token, expires } = data;

        if (origin !== HOST || !token) {
          return;
        }

        window.localStorage.setItem(TOKEN_KEY, token);
        window.localStorage.setItem(TOKEN_EXPIRATION, expires);
        this.setState({ client: this.makeClient(data.token), accessToken: token, expires });
      },
      false
    );
  }

  makeClient = token => {
    return new OptimizelyClient({
      accessToken: token,
      project: this.props.sdk.parameters.installation.optimizelyProjectId,
      onReauth: () => {
        this.setState({ client: null });
      }
    });
  };

  openAuth = () => {
    const WINDOW_OPTS = 'left=150,top=150,width=700,height=700';
    window.open(url, '', WINDOW_OPTS);
  };

  render() {
    const { state, props } = this;
    const { sdk } = props;
    const { client } = state;
    const { location, parameters } = sdk;

    if (location.is(locations.LOCATION_APP)) {
      return (
        <AppPage
          openAuth={this.openAuth}
          accessToken={this.state.accessToken}
          sdk={this.props.sdk}
          client={this.state.client}
        />
      );
    }

    if (location.is(locations.LOCATION_ENTRY_SIDEBAR)) {
      if (!parameters.installation.optimizelyProjectId) {
        return <MissingProjectId />;
      }

      return <Sidebar sdk={sdk} />;
    }

    if (location.is(locations.LOCATION_ENTRY_EDITOR)) {
      const [valid, missingFields] = isValidContentType(sdk.contentType);

      if (!valid) {
        return <IncorrectContentType sdk={sdk} missingFields={missingFields} />;
      }

      return (
        <EditorPage
          sdk={sdk}
          client={client}
          openAuth={this.openAuth}
          expires={this.state.expires}
        />
      );
    }
  }
}

init(sdk => {
  render(<App sdk={sdk} />, document.getElementById('root'));
});
