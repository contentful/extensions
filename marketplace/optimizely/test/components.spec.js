import React from 'react';
import { render, cleanup, configure } from '@testing-library/react';

import { IncorrectContentType, isValidContentType, MissingProjectId } from '../src/errors-messages';

configure({ testIdAttribute: 'data-test-id' });

describe('error message components', () => {
  afterEach(cleanup);
  describe('IncorrectContentType', () => {
    it('should match snapshot', () => {
      const mockSdk = {
        ids: {
          extension: 'test-extension'
        }
      };

      const missingFields = [{ id: 'key', type: 'symbol' }];

      const { container } = render(
        <IncorrectContentType sdk={mockSdk} missingFields={missingFields} />
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('MissingProjectId', () => {
    it('should match snapshot', () => {
      const { container } = render(<MissingProjectId />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('isValidContentType', () => {
    const makeFields = (valid = false) => {
      return {
        fields: [
          {
            id: 'experimentTitle',
            name: 'Experiment title',
            type: 'Symbol',
            localized: false,
            required: false,
            validations: [],
            disabled: false,
            omitted: false
          },
          {
            id: 'experimentId',
            name: 'Experiment ID',
            type: 'Symbol',
            localized: false,
            required: false,
            validations: [],
            disabled: false,
            omitted: false
          },
          {
            id: 'meta',
            name: 'Meta',
            type: 'Object',
            localized: false,
            required: false,
            validations: [],
            disabled: false,
            omitted: false
          },
          {
            id: 'variations',
            name: 'Variations',
            type: 'Array',
            localized: false,
            required: false,
            validations: [],
            disabled: false,
            omitted: false,
            items: {
              type: 'Link',
              validations: [],
              linkType: 'Entry'
            }
          }
        ].concat(
          valid
            ? {
                id: 'experimentKey',
                name: 'Experiment key',
                type: 'Symbol',
                localized: false,
                required: false,
                validations: [],
                disabled: false,
                omitted: false
              }
            : []
        )
      };
    };
    it('should produce an invalid response', () => {
      expect(isValidContentType(makeFields())).toEqual([
        false,
        [{ id: 'experimentKey', type: 'Symbol' }]
      ]);
    });
    it('should produce a valid response', () => {
      expect(isValidContentType(makeFields(true))).toEqual([true, []]);
    });
  });
});
