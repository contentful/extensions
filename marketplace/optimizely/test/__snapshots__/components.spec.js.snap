// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error message components IncorrectContentType should match snapshot 1`] = `
<div>
  <div
    class="css-hvj582"
    data-test-id="incorrect-type"
  >
    <div
      class="Note__Note___2eSKN Note__Note--negative___3fkwL"
      data-test-id="cf-ui-note"
    >
      <div
        class="Note__Note__icon___20RqC"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--negative___1dled"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
      </div>
      <div>
        <div
          class="Note__Note__title___2nwpx"
        >
          Extension "test-extension" cannot be assigned to this content type
        </div>
        <div>
          <div>
            <strong>
              Required:
            </strong>
             
            experimentTitle (Symbol), experimentId (Symbol), experimentKey (Symbol), meta (Object), variations (Array)
          </div>
          <div>
            <strong>
              Missing:
            </strong>
             
            key (symbol)
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`error message components MissingProjectId should match snapshot 1`] = `
<div>
  <div
    class="css-hvj582"
    data-test-id="missing-project"
  >
    <div
      class="Note__Note___2eSKN Note__Note--negative___3fkwL"
      data-test-id="cf-ui-note"
    >
      <div
        class="Note__Note__icon___20RqC"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--negative___1dled"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
      </div>
      <div>
        <div>
          Please, set 
          <strong>
            Optimizely Project ID
          </strong>
           in Extension settings.
        </div>
      </div>
    </div>
  </div>
</div>
`;
